### PR TITLE
ARROW-10263: [C++][Compute] Improve variance kernel numerical stability

### PR DESCRIPTION
**Pull request by cyb70289**
Improve variance merging method to address stability issue when merging
short chunks with approximate mean value.

Improve reference variance accuracy by leveraging Kahan summation.

### CODE REVIEW

**Comment by cyb70289** [comment]
CI failure looks not related

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 1]
```diff
@@ -1070,22 +1065,39 @@ TEST_F(TestVarStdKernelStability, Basics) {
   VarianceOptions options{1};  // ddof = 1
   this->AssertVarStdIs("[100000004, 100000007, 100000013, 100000016]", options, 30.0);
   this->AssertVarStdIs("[1000000004, 1000000007, 1000000013, 1000000016]", options, 30.0);
+
+#ifndef __MINGW32__  // MinGW has precision issues
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:1069]: This was only the 32-bit MinGW build, i.e. it was perhaps not MinGW but x87 (perhaps you can check with a 32-bit Linux build?).

Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1069]: This test failed on mingw 32 community CI. And I see similar comments in decimal unit test.
https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/decimal_test.cc#L695

I didn't tested it on my side. Maybe I can start a 32bit VM to check.

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1069]: Ok, I've checked and there is no failure on Linux i386. It does seem MinGW-related.

**Review by pitrou — APPROVED** [review]
Will merge. Thanks a lot for doing this!

**Comment by alippai** [comment]
Are there any before/after benchmarks? It's really nice that we can have extra numerical stability, I'm just curious what's the penalty for it.

**Comment by cyb70289** [comment]
> Are there any before/after benchmarks? It's really nice that we can have extra numerical stability, I'm just curious what's the penalty for it.

This change is only for combing variances from multiple arrays. The time is trivial compared with computing variance for each array.
Benchmark also shows no difference (benchmark PR is pending review, https://github.com/apache/arrow/pull/8407)

**Comment by alippai** [comment]
Amazing, thanks!

**Pull request closed without merging**